Add lint to detect `allow` attributes without reason

xFrednet opened the pull request:
I was considering putting this lint into the pedantic group. However, that would result in countless warnings for existing projects. Having it in restriction also seems good to me :upside_down_face: (And now I need sleep :zzz: )

---

changelog: New lint [`allow_lint_without_reason`] (Requires the `lint_reasons` feature)

Closes: rust-lang/rust-clippy#8502

flip1995 commented on clippy_lints/src/attrs.rs:
````diff
@@ -255,7 +255,38 @@ declare_clippy_lint! {
+    /// #![allow(clippy::some_lint, reason = "False positive rust-lang/rust-clippy#1002020")]
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub ALLOW_LINT_WITHOUT_REASON,
````
Lint name suggestion:
```suggestion
    pub ALLOW_ATTRIBUTES_WITHOUT_REASON,
```
This flows better with `allow(clippy::allow_attributes_without_reason)`. Still not perfect, but better than with the current lint name IMO

flip1995 commented on clippy_lints/src/attrs.rs:
```diff
@@ -404,6 +438,30 @@ fn check_clippy_lint_names(cx: &LateContext<'_>, name: Symbol, items: &[NestedMe
+        cx,
+        ALLOW_LINT_WITHOUT_REASON,
+        attr.span,
+        &format!("`{}` attribute without reason", name.as_str()),
```
```suggestion
        &format!("`{}` attribute without specifying a reason", name.as_str()),
```

maybe?

flip1995 reviewed (commented):
LGTM overall. `restriction` sounds like the orrect lint group to me :+1:

xFrednet commented:
Good suggestions :+1: The lint name was also something I was thinking about :upside_down_face: I've added your suggestions and squashed the commits :upside_down_face:

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit ab6ffb6371b78d1777b90c25a44163562ae53036 has been approved by `flip1995`

<!-- @bors r=flip1995 ab6ffb6371b78d1777b90c25a44163562ae53036 -->
<!-- homu: {"type":"Approved","sha":"ab6ffb6371b78d1777b90c25a44163562ae53036","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit ab6ffb6371b78d1777b90c25a44163562ae53036 with merge 7ae527faa3c5393eeedd88117560da08dc75a719...
<!-- homu: {"type":"BuildStarted","head_sha":"ab6ffb6371b78d1777b90c25a44163562ae53036","merge_sha":"7ae527faa3c5393eeedd88117560da08dc75a719"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/5427116886?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/5427116886?check_suite_focus=true","builder_name":"checks-action_test"} -->

matthiaskrgr commented:
@bors retry
````

Run actions/download-artifact@v1
Downloading artifact 'target' to: '/home/runner/work/rust-clippy/rust-clippy/target'
Error: The HTTP request timed out after 00:01:40.
Error: Exit code 1 returned from process: file name '/home/runner/runners/2.288.1/bin/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.DownloadArtifact, Runner.Plugins"'.
````

bors commented:
:hourglass: Testing commit ab6ffb6371b78d1777b90c25a44163562ae53036 with merge 48d54942f55078cd5866a96bd182bbaf2b81d0f6...
<!-- homu: {"type":"BuildStarted","head_sha":"ab6ffb6371b78d1777b90c25a44163562ae53036","merge_sha":"48d54942f55078cd5866a96bd182bbaf2b81d0f6"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/5427163127?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/5427150058?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/5427277255?check_suite_focus=true)
Approved by: flip1995
Pushing 48d54942f55078cd5866a96bd182bbaf2b81d0f6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/5427277255?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/5427163127?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/5427150058?check_suite_focus=true"},"merge_sha":"48d54942f55078cd5866a96bd182bbaf2b81d0f6"} -->
